pkg/deviceclaimingserver/grpc_end_devices.go

```diff
@@ -86,15 +93,67 @@ type endDeviceClaimingServer struct {
 
 // Claim implements EndDeviceClaimingServer.
 func (edcs *endDeviceClaimingServer) Claim(ctx context.Context, req *ttnpb.ClaimEndDeviceRequest) (*ttnpb.EndDeviceIdentifiers, error) {
-	ids, err := edcs.DCS.endDeviceClaimingUpstream.Claim(ctx, req)
+	// Check that the collaborator has necessary rights before attempting to claim it on an upstream.
+	// Since this is part of the create device flow, we check that the collaborator has the rights to create devices in the application.
+	targetAppID := req.GetTargetApplicationIds()
+	if err := rights.RequireApplication(ctx, *targetAppID,
+		ttnpb.Right_RIGHT_APPLICATION_DEVICES_WRITE,
+	); err != nil {
+		return nil, err
+	}
+
+	var (
+		joinEUI                 *types.EUI64
+		devEUI                  *types.EUI64
+		claimAuthenticationCode string
+	)
+	if authenticatedIDs := req.GetAuthenticatedIdentifiers(); authenticatedIDs != nil {
+		joinEUI = &req.GetAuthenticatedIdentifiers().JoinEui
+		devEUI = &req.GetAuthenticatedIdentifiers().DevEui
+		claimAuthenticationCode = req.GetAuthenticatedIdentifiers().AuthenticationCode
+	} else if qrCode := req.GetQrCode(); qrCode != nil {
+		conn, err := edcs.DCS.GetPeerConn(ctx, ttnpb.ClusterRole_QR_CODE_GENERATOR, nil)
+		if err != nil {
+			return nil, err
+		}
+		qrg := ttnpb.NewEndDeviceQRCodeGeneratorClient(conn)
+		data, err := qrg.Parse(ctx, &ttnpb.ParseEndDeviceQRCodeRequest{
+			QrCode: qrCode,
+		}, edcs.DCS.WithClusterAuth())
+		dev := data.GetEndDeviceTemplate().GetEndDevice()
+		if dev == nil {
+			return nil, errParseQRCode.New()
+		}
+		joinEUI = dev.GetIds().JoinEui
+		devEUI = dev.GetIds().DevEui
+		claimAuthenticationCode = dev.ClaimAuthenticationCode.Value
+	} else {
+		return nil, errNoJoinEUI.New()
+	}
+
+	networkServerAddress, _, err := net.SplitHostPort(req.TargetNetworkServerAddress)
+	if err != nil {
+		// TargetNetworkServerAddress is already validated by the API.
+		// An error here means that it does not contain a port, so we use it directly.
+		networkServerAddress = req.TargetNetworkServerAddress
+	}
+
+	err = edcs.DCS.endDeviceClaimingUpstream.Claim(ctx, joinEUI, devEUI, claimAuthenticationCode, networkServerAddress)
```

Comment:
Please put a comment here for our future selves why we're taking the host here and not the address.

Like:

> External services, including Join Servers, typically identify Network Servers by host instead of by host and port.